### PR TITLE
Support Azure DNS and VXLAN node network policy

### PR DESCRIPTION
**Pull request by vomba**
> [!warning]
> **This is a public repository, ensure not to disclose:**
>
> - [x] personal data beyond what is necessary for interacting with this pull request, nor
> - [x] business confidential information, such as customer names.

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [x] kind/feature       
- [ ] kind/improvement   
- [ ] kind/deprecation   
- [ ] kind/documentation 
- [ ] kind/clean-up      
- [ ] kind/bug           
- [ ] kind/other         

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change 
- [ ] kind/dev-change   
- [ ] kind/security     
- [ ] [kind/adr]()      

### What does this PR do / why do we need this PR?

Adds support to initialize config with support to CAPI and IPVXLAN on azure

- Adds one #1999 

#### Checklist

- [ ] Proper commit message prefix on all commits
  
- Change checks:
  - [x] The change is transparent
  - [ ] The change is disruptive
  - [ ] The change requires no migration steps
  - [ ] The change requires migration steps
  - [ ] The change upgrades CRDs
- Metrics checks:
  - [ ] The metrics are still exposed and present in Grafana after the change
  - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [ ] The logs do not show any errors after the change
- Pod Security Policy checks:
  - [ ] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [ ] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Network Policy checks:
  - [ ] Any changed pod is covered by Network Policies
  - [ ] The change does not cause any dropped packets in the `NetworkPolicy Dashboard`
- Audit checks:
  - [ ] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy
- Falco checks:
  - [ ] The change does not cause any alerts to be generated by Falco
- Bug checks:
  - [ ] The bug fix is covered by regression tests

### CODE REVIEW

**Comment by aarnq** [comment]
Since this isn't dependent on your storage resource script you should rebase it on main and remove all commits except the new ones.

File: bin/init.bash [hunk 1]
```diff
@@ -6,6 +6,7 @@
 # It's not to be executed on its own but rather via `ck8s init`.
 
 : "${CK8S_CLUSTER:?Missing CK8S_CLUSTER}"
+: "${CLUSTER_PROVISIONER:?Missing CLUSTER_PROVISIONER}"
```
Comment (aarnq) [inline bin/init.bash:9]: Please add this as a variable comparable `CK8S_ENVIRONMENT_NAME`/`CK8S_FLAVOR`/`CK8S_CLOUD_PROVIDER` in common.bash:
```diff
diff --git a/bin/common.bash b/bin/common.bash
index 11c83ebd..5c6a862f 100644
--- a/bin/common.bash
+++ b/bin/common.bash
@@ -9,6 +9,7 @@
 # shellcheck disable=SC2034
 ck8s_cloud_providers=(
     "aws"
+    "azure"
     "baremetal"
     "citycloud"
     "exoscale"
@@ -18,6 +19,13 @@ ck8s_cloud_providers=(
     "none"
 )
 
+# shellcheck disable=SC2034
+ck8s_cluster_provisioner=(
+    "cluster-api"
+    "kubespray"
+    "other"
+)
+
 # shellcheck disable=SC2034
 ck8s_flavors=(
     "dev"
```

Then in init to support it:
```diff
diff --git a/bin/init.bash b/bin/init.bash
index e09d79b6..b98af2b4 100755
--- a/bin/init.bash
+++ b/bin/init.bash
@@ -18,6 +18,7 @@ source "${here}/common.bash"
 # Load cloud provider, environment name, and flavor from config if available.
 if [ -f "${config[default_common]}" ]; then
     cloud_provider=$(yq4 '.global.ck8sCloudProvider' "${config[default_common]}")
+    cluster_provisioner=$(yq4 '.global.ck8sClusterProvisioner' "${config[default_common]}")
     environment_name=$(yq4 '.global.ck8sEnvironmentName' "${config[default_common]}")
     flavor=$(yq4 '.global.ck8sFlavor' "${config[default_common]}")
 fi
@@ -29,6 +30,14 @@ elif [ -v CK8S_CLOUD_PROVIDER ] && [ "${CK8S_CLOUD_PROVIDER}" != "${cloud_provid
 else
     export CK8S_CLOUD_PROVIDER="${cloud_provider}"
 fi
+if [ -z "${cluster_provisioner:-}" ]; then
+    : "${CK8S_CLUSTER_PROVISIONER:?Missing CK8S_CLUSTER_PROVISIONER}"
+elif [ -v CK8S_CLUSTER_PROVISIONER ] && [ "${CK8S_CLUSTER_PROVISIONER}" != "${cluster_provisioner}" ]; then
+    log_error "ERROR: Cloud provider mismatch, '${cluster_provisioner}' in config and '${CK8S_CLUSTER_PROVISIONER}' in env"
+    exit 1
+else
+    export CK8S_CLOUD_PROVIDER="${cloud_provider}"
+fi
 if [ -z "${environment_name:-}" ]; then
     : "${CK8S_ENVIRONMENT_NAME:?Missing CK8S_ENVIRONMENT_NAME}"
 elif [ -v CK8S_ENVIRONMENT_NAME ] && [ "${CK8S_ENVIRONMENT_NAME}" != "${environment_name}" ]; then
@@ -410,16 +419,21 @@ set_cluster_dns() {
         exit 1
     fi
     case ${CLUSTER_PROVISIONER} in
-        CAPI)
+        cluster-api)
             clusterapi=true
             clusterdns="10.233.0.10"
             clusterdnscidr="10.233.0.10/32"
         ;;
-        *)
+        kubespray)
             clusterapi=false
             clusterdns="10.233.0.3"
             clusterdnscidr="10.233.0.3/32"
         ;;
+        *)
+            clusterapi=false
+            clusterdns="set-me"
+            clusterdnscidr="set-me"
+        ;;
     esac
 
     replace_set_me "${file}" ".global.clusterDns" "\"${clusterdns}\""
```

This will also require migration steps to set the correct one for current clusters.

**Review by raviranjanelastisys — APPROVED** [review]

File: bin/init.bash [hunk 2]
```diff
@@ -396,6 +411,38 @@ set_harbor_config() {
     replace_set_me "${file}" ".harbor.persistence.disableRedirect" "${disable_redirect}"
 }
 
+# Usage: set_cluster_dns <config-file>
+set_cluster_dns() {
```
Comment (simonklb) [inline bin/init.bash:415]: This does more than just configuring DNS. Could it be named something more generic to setting "cluster provisioner specific configuration options" or could it be split into `set_cluster_dns` and `set_cluster_api`?

File: bin/update-ips.bash [hunk 3]
```diff
@@ -20,6 +20,9 @@ if [[ "${2}" == "apply" ]]; then
 fi
 has_diff=0
 
+#TODO: To be changed when decision made on networkpolicies for azure storage
+STORAGE_SERVICE=$(yq4 '.objectStorage.type' "${CK8S_CONFIG_PATH}/defaults/common-config.yaml")
```
Comment (simonklb) [inline bin/update-ips.bash:24]: Make this lowercase to align with the rest of the script. Uppercase variables are primarily used to indicate variables that are in your environment.

**Review by simonklb — CHANGES_REQUESTED** [review]

**Review by simonklb — CHANGES_REQUESTED** [review]
This is backwards incompatible. It should be mentioned as an ops note and the PR should include migration steps/script for existing environments.

Edit: Looks like some tests also needs to be updated!

File: bin/init.bash (hunk 1 above)
Comment (aarnq) [inline bin/init.bash:9]:
```suggestion
```
We will use `CK8S_CLUSTER_PROVISIONER` instead, which is loaded in common (if it is set).

File: bin/init.bash [hunk 4]
```diff
@@ -396,6 +411,56 @@ set_harbor_config() {
...
 
+
+# Usage: set_cluster_api <config-file>
+set_cluster_api() {
+    file="${1}"
+    if [[ ! -f "${file}" ]]; then
+        log_error "ERROR: invalid file - ${file}"
+        exit 1
+    fi
+    case ${CLUSTER_PROVISIONER} in
```
Comment (aarnq) [inline bin/init.bash:422]:
```suggestion
    case ${CK8S_CLUSTER_PROVISIONER} in
```

File: bin/init.bash [hunk 5]
```diff
@@ -396,6 +411,56 @@ set_harbor_config() {
...
+}
+
+# Usage: set_cluster_dns <config-file>
+set_cluster_dns() {
+    file="${1}"
+    if [[ ! -f "${file}" ]]; then
+        log_error "ERROR: invalid file - ${file}"
+        exit 1
+    fi
+    case ${CLUSTER_PROVISIONER} in
```
Comment (aarnq) [inline bin/init.bash:445]:
```suggestion
    case ${CK8S_CLUSTER_PROVISIONER} in
```

**Review by aarnq — COMMENTED** [review]
For the test you need to add [here](https://github.com/elastisys/compliantkubernetes-apps/blob/main/tests/common/lib/env.bash#L6-L12): `export CK8S_CLUSTER_PROVISIONER="kubespray"`

(We can add "cluster-api" as a test target later.)

You also need to check other `.bats` files that doesn't use `env.setup / env.init` that they export `CK8S_CLUSTER_PROVISIONER="kubespray"` when they set variables.
Example is [this one](https://github.com/elastisys/compliantkubernetes-apps/blob/main/tests/unit/bin/init.bats#L17-L27), just check which bats tests fail and they should print the file.

**Comment by vomba** [comment]
Removed the need for CLUSTER_PROVISIONER and set the related settings to be azure related.

File: bin/common.bash [hunk 6]
```diff
@@ -18,6 +19,13 @@ ck8s_cloud_providers=(
     "none"
 )
 
+# shellcheck disable=SC2034
+ck8s_cluster_provisioner=(
+    "cluster-api"
+    "kubespray"
+    "other"
+)
+
```
Comment (simonklb) [inline bin/common.bash:28]:
```suggestion
```

File: bin/init.bash [hunk 7]
```diff
@@ -15,11 +15,6 @@ here="$(dirname "$(readlink -f "$0")")"
 source "${here}/common.bash"
 
 # Load cloud provider, environment name, and flavor from config if available.
-if [ -f "${config[default_common]}" ]; then
-    cloud_provider=$(yq4 '.global.ck8sCloudProvider' "${config[default_common]}")
-    environment_name=$(yq4 '.global.ck8sEnvironmentName' "${config[default_common]}")
-    flavor=$(yq4 '.global.ck8sFlavor' "${config[default_common]}")
-fi
```
Comment (simonklb) [inline bin/init.bash:22]: Why is this removed?

**Review by simonklb — CHANGES_REQUESTED** [review]

File: bin/init.bash (hunk 7 above)
Comment (vomba) [reply bin/init.bash:22]: By mistake will add it back.

**Review by simonklb — APPROVED** [review]
:partying_face:

File: tests/common/lib/update-ips.bash [hunk 8]
```diff
@@ -41,16 +41,20 @@ update_ips.mock_minimal() {
...
+  mock_set_output "${mock_kubectl}" "127.0.1.9 127.0.2.9 127.0.3.9" 8 # .networkPolicies.global.scNodes.ips calico wireguard
+  mock_set_output "${mock_kubectl}" "127.0.1.4 127.0.2.4 127.0.3.4" 9 # .networkPolicies.global.wcApiserver.ips node internal
+  mock_set_output "${mock_kubectl}" "127.0.1.5 127.0.2.5 127.0.3.5" 10 # .networkPolicies.global.wcApiserver.ips calico ipip
+  mock_set_output "${mock_kubectl}" "127.0.1.5 127.0.2.5 127.0.3.5" 11 # .networkPolicies.global.wcApiserver.ips calico vxlan
+  mock_set_output "${mock_kubectl}" "127.0.1.6 127.0.2.6 127.0.3.6" 12 # .networkPolicies.global.wcApiserver.ips calico wireguard
+  mock_set_output "${mock_kubectl}" "127.0.1.10 127.0.2.10 127.0.3.10" 13 # .networkPolicies.global.wcNodes.ips node internal
+  mock_set_output "${mock_kubectl}" "127.0.1.11 127.0.2.11 127.0.3.11" 14 # .networkPolicies.global.wcNodes.ips calico ipip
+  mock_set_output "${mock_kubectl}" "127.0.1.11 127.0.2.11 127.0.3.11" 15 # .networkPolicies.global.wcNodes.ips calico vxlan
+  mock_set_output "${mock_kubectl}" "127.0.1.12 127.0.2.12 127.0.3.12" 16 # .networkPolicies.global.wcNodes.ips calico wireguard
```
Comment (simonklb) [inline tests/common/lib/update-ips.bash:57]: Using the same IPs as the mocked output for vxlan and ipip lookups is not going to test it properly. If you excluded the results of either of of them in the final network policy configuration this would still pass. Please use different IPs for each of them.

**Review by simonklb — CHANGES_REQUESTED** [review]

**Review by simonklb — APPROVED** [review]
:rocket:

**Comment by vomba** [comment]
> Please give the PR a proper title btw

Title edited.

**Comment by simonklb** [comment]
> It basically adds support for ck8s-cluster-api dns IPs, IPVXLAN, and azure. (IPVXLAN is needed for azure but this makes it not azure specific)

Right, what about "Support Azure DNS and VXLAN node network policy" or something like that?

**Comment by simonklb** [comment]
Either way, this is fine to merge whenever you're ready! :)
